src/Features/LanguageServer/Protocol/Features/EditAndContinue/EditAndContinueDiagnosticSource.cs

```diff
@@ -0,0 +1,140 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
+using Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.EditAndContinue;
+
+internal static class EditAndContinueDiagnosticSource
+{
+    private sealed class ProjectSource(Project project, ImmutableArray<DiagnosticData> diagnostics) : AbstractProjectDiagnosticSource(project)
```

Comment:
interesting.  i don't mind this (that you capture the diags at creation point and then just return them.  But i'm curious why do it this way, vs actually having GetDiagnosticsAsync call into your EnC service to fetch them when requested?  that seems more 'morally right' in terms of how the flow is supposed to go.

Reply:
These diagnostics require explicit user action of applying an edit. We don't have GetDiagnosticsAsync that would fetch them. They are calculated during EmitUpdate.